src/Adapter/MSTest.TestAdapter/Execution/ClassCleanupManager.cs

```diff
@@ -62,4 +62,19 @@ public void MarkTestComplete(TestMethodInfo testMethodInfo, TestMethod testMetho
             ShouldRunEndOfAssemblyCleanup = _remainingTestsByClass.IsEmpty;
         }
     }
+
+    internal static void ForceCleanup(TypeCache typeCache)
+    {
+        IEnumerable<TestClassInfo> classInfoCache = typeCache.ClassInfoListWithExecutableCleanupMethods;
```

Comment:
If init wasn't called, then we will not call cleanup. This is handled in ExecuteClassCleanup by this piece of code:

```
if (IsClassCleanupExecuted
    // If there is a ClassInitialize method and it has not been executed, then we should not execute ClassCleanup
    || (!IsClassInitializeExecuted && ClassInitializeMethod is not null))
{
    return;
}
```

However, thinking now more, it's curious what will happen if there is only class cleanup but there is no class initialize.
Generally, if the class didn't yet start executing, it's not in cache, but it's definitely not a good idea to rely on that behavior.

The logic I mentioned above (which is already existing) feels somewhat suspicious to me as well in case the class doesn't define ClassCleanup, but base class does.
It doesn't make sense we check whether ClassInitializeMethod is null or not without also a corresponding check for base classes.

Reply:
Adjusted the logic there a bit. It should be better now.